src/runtime/process.rs

```diff
@@ -566,29 +568,75 @@ impl Default for LinuxSchedulerFlag {
     default,
     pattern = "owned",
     setter(into, strip_option),
-    build_fn(error = "OciSpecError")
+    build_fn(validate = "Self::validate", error = "OciSpecError")
 )]
 #[getset(get = "pub", set = "pub")]
 /// ExecCPUAffinity specifies CPU affinity used to execute the process.
 /// This setting is not applicable to the container's init process.
 pub struct ExecCPUAffinity {
-    #[serde(default, skip_serializing_if = "Option::is_none")]
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        deserialize_with = "deserialize"
+    )]
     /// cpu_affinity_initial is a list of CPUs a runtime parent process to be run on
     /// initially, before the transition to container's cgroup.
     /// This is a a comma-separated list, with dashes to represent ranges.
     /// For example, `0-3,7` represents CPUs 0,1,2,3, and 7.
     cpu_affinity_initial: Option<String>,
 
-    #[serde(default, skip_serializing_if = "Option::is_none")]
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        deserialize_with = "deserialize"
+    )]
     /// cpu_affinity_final is a list of CPUs the process will be run on after the transition
     /// to container's cgroup. The format is the same as for `initial`. If omitted or empty,
     /// the container's default CPU affinity, as defined by cpu.cpus property, is used.
     cpu_affinity_final: Option<String>,
 }
 
+impl ExecCPUAffinityBuilder {
+    fn validate(&self) -> Result<(), OciSpecError> {
+        if let Some(Some(ref s)) = self.cpu_affinity_initial {
+            validate_cpu_affinity(s).map_err(|e| OciSpecError::Other(e.to_string()))?;
+        }
+
+        if let Some(Some(ref s)) = self.cpu_affinity_final {
+            validate_cpu_affinity(s).map_err(|e| OciSpecError::Other(e.to_string()))?;
+        }
+
+        Ok(())
+    }
+}
+
+fn deserialize<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let value: Option<String> = Option::deserialize(deserializer)?;
+
+    if let Some(ref s) = value {
+        validate_cpu_affinity(s).map_err(de::Error::custom)?;
+    }
+
+    Ok(value)
+}
+
+fn validate_cpu_affinity(s: &str) -> Result<(), String> {
+    let re = Regex::new(r"^(\d+(-\d+)?)(,\d+(-\d+)?)*$").map_err(|e| e.to_string())?;
```

Comment:
What do folks think about using once_cell or lazy_static for this regex to compile it only once?

Reply:
Hi @keisku @saschagrunert I want to confirm that using regex to  do such work expensive or not ?